src/devices/bus/multibus/robotron_k7070.cpp

```diff
@@ -0,0 +1,408 @@
+	map(0x8000, 0xffff).view(m_view_hi);
+	m_view_hi[0](0x8000, 0xffff).ram().share("kgs_ram2");
+	m_view_hi[1](0x8000, 0xffff).ram().share("vram0");
+}
+
+void robotron_k7070_device::cpu_pio(address_map &map)
+{
+	map.unmap_value_high();
+	map.global_mask(0xff);
+
+	map(0x0000, 0x0003).rw(m_ctc, FUNC(z80ctc_device::read), FUNC(z80ctc_device::write));
+	map(0x0008, 0x000b).rw(m_sio, FUNC(z80sio_device::ba_cd_r), FUNC(z80sio_device::ba_cd_w));
+	map(0x0010, 0x0017).rw(FUNC(robotron_k7070_device::kgs_host_r), FUNC(robotron_k7070_device::kgs_host_w)); // p. 11 of KGS-K7070.pdf
+
+	// p. 6 of ABG-K7072.pdf
+	map(0x0020, 0x0021).w(FUNC(robotron_k7070_device::abg_addr_w));
+	map(0x0022, 0x0022).w(FUNC(robotron_k7070_device::abg_func_w));
+	map(0x0023, 0x0023).w(FUNC(robotron_k7070_device::abg_split_w));
+	map(0x0030, 0x003f).noprw(); // palette register
+
+	map(0x0080, 0x0080).w(FUNC(robotron_k7070_device::abg_misc_w));
+}
+
+void robotron_k7070_device::palette_init(palette_device &palette)
+{
+	// 0=black, 1=dim, 2=normal, 3=bold
+
+	palette.set_pen_color(0, rgb_t::black());
+	palette.set_pen_color(1, rgb_t(0x80, 0x80, 0x80));
+	palette.set_pen_color(2, rgb_t(0xc0, 0xc0, 0xc0));
+	palette.set_pen_color(3, rgb_t::white());
+}
+
+uint16_t robotron_k7070_device::io_r(offs_t offset, uint16_t mem_mask)
+{
+	uint8_t data = 0;
+
+	switch (offset)
+	{
+	case 0:
+		data = m_kgs_ctrl;
+		break;
+
+	case 1:
+		data = m_kgs_datai;
+		m_kgs_ctrl &= ~KGS_ST_OBF;
+		break;
+	}
+
+	if (offset)
+	LOG("%s: KGS %d == %02x '%c'\n", machine().describe_context(), offset, data,
+			 (data > 0x1f && data < 0x7f) ? data : 0x20);
+
+	return data;
+}
+
+void robotron_k7070_device::io_w(offs_t offset, uint16_t data, uint16_t mem_mask)
+{
+	if (offset)
+	LOG("%s: KGS %d <- %02x '%c', ctrl %02x\n", machine().describe_context(), offset, data,
+			 (data > 0x1f && data < 0x7f) ? data : 0x20, m_kgs_ctrl);
+
+	switch (offset)
+	{
+	case 0:
+		m_kgs_ctrl &= ~(KGS_ST_ERR | KGS_ST_INT);
+		int_w(7, CLEAR_LINE);
+		break;
+
+	case 1:
+		if (m_kgs_ctrl & KGS_ST_IBF)
+		{
+			m_kgs_ctrl |= KGS_ST_ERR;
+		}
+		else
+		{
+			m_kgs_datao = data;
+			m_kgs_ctrl |= KGS_ST_IBF;
+		}
+		break;
+	}
+}
+
+uint32_t robotron_k7070_device::screen_update_k7072(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect)
+{
+	int addr = m_start;
+
+	for (int y = 0; y < 400; y++)
+	{
+		int horpos = 0;
+		for (int x = 0; x < 80; x++)
+		{
+			uint8_t code = m_ram[addr++ % 32768];
+			for (int b = 0; b < 8; b++)
+			{
+				bitmap.pix(y, horpos++) = (code >> (7 - b)) & 1;
+			}
+		}
+	}
+
+	return 0;
+}
+
+
+uint8_t robotron_k7070_device::kgs_host_r(offs_t offset)
+{
+	uint8_t data = 0;
+
+	switch (offset)
+	{
+	case 0:
+		data = m_kgs_datao;
+		m_kgs_ctrl &= ~(KGS_ST_ERR | KGS_ST_IBF);
+		break;
+
+	case 2:
+		data = m_kgs_ctrl;
+		break;
+
+	case 6:
+		data = ioport("DSEL0")->read();
+		break;
+
+	case 7:
+		data = ioport("DSEL1")->read();
+		break;
```

Comment:
Please use a `required_ioport_array<2>` to access these so they don’t cause tag lookups in a read handler.